iroh-gateway/src/config.rs

```diff
@@ -2,122 +2,167 @@ use std::net::SocketAddr;
 
 use crate::constants::*;
 use axum::http::{header::*, Method};
+use config::{ConfigError, Map, Source, Value};
 use headers::{
     AccessControlAllowHeaders, AccessControlAllowMethods, AccessControlAllowOrigin, HeaderMapExt,
 };
-use iroh_rpc_client::RpcClientConfig;
+use iroh_rpc_client::Config as RpcClientConfig;
+use iroh_util::insert_into_config_map;
+use serde::{Deserialize, Serialize};
 
+/// CONFIG_FILE_NAME is the name of the optional config file located in the iroh home directory
+pub const CONFIG_FILE_NAME: &str = "gateway.config.toml";
+/// ENV_PREFIX should be used along side the config field name to set a config field using
+/// environment variables
+/// For example, `IROH_GATEWAY_PORT=1000` would set the value of the `Config.port` field
+pub const ENV_PREFIX: &str = "IROH_GATEWAY";
 pub const DEFAULT_PORT: u16 = 9050;
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
 pub struct Config {
     /// flag to toggle whether the gateway allows writing/pushing data
     pub writeable: bool,
     /// flag to toggle whether the gateway allows fetching data from other nodes or is local only
     pub fetch: bool,
     /// flag to toggle whether the gateway enables/utilizes caching
     pub cache: bool,
-    /// set of user provided headers to attach to all responses
-    pub headers: HeaderMap,
     /// default port to listen on
     pub port: u16,
-    pub rpc: RpcConfig,
-}
-
-#[derive(Debug, Clone)]
-pub struct RpcConfig {
-    /// Address on which to listen,
-    pub listen_addr: SocketAddr,
-    pub client_config: RpcClientConfig,
-}
-
-impl Default for RpcConfig {
-    fn default() -> Self {
-        let client_config = RpcClientConfig::default();
-        RpcConfig {
-            listen_addr: client_config.gateway_addr,
-            client_config,
-        }
-    }
+    /// rpc listening addr
+    pub rpc_addr: SocketAddr,
```

Comment:
changed this not because I necessarily think it's better, but just so that the way we are handling the rpc_addr and rpc_clients is uniform over each config.